### PR TITLE
(SIMP-4798) Improved reboot_notify error handling

### PR DESCRIPTION
**Pull request by trevor-vaughan**
SIMP-4799 #close

### CODE REVIEW

File: lib/puppet/provider/reboot_notify/notify.rb [hunk 1]
```diff
@@ -36,7 +40,11 @@ def update
       :updated => Time.now.tv_sec
     }
 
-    File.open(@target,'w'){|fh| fh.puts(PSON.pretty_generate(@records))}
+    begin
+      File.open(@target,'w'){|fh| fh.puts(JSON.pretty_generate(@records))}
+    rescue
```
Comment (lnemsick-simp) [inline lib/puppet/provider/reboot_notify/notify.rb]: See comment above about rescuing specific exceptions and including the specifics of that error in the error message.

File: lib/puppet/provider/reboot_notify/notify.rb [hunk 2]
```diff
@@ -57,18 +65,22 @@ def self.post_resource_eval
...
         msg << ["  #{k} => #{v['reason']}"]
       end
 
       Puppet.notice(msg.join("\n"))
     end
 
-    File.open(target,'w'){|fh| fh.puts(PSON.pretty_generate(records))}
+    begin
+      File.open(target,'w'){|fh| fh.puts(JSON.pretty_generate(records))}
+    rescue
```
Comment (lnemsick-simp) [inline lib/puppet/provider/reboot_notify/notify.rb]: See comment above about rescuing specific exceptions and including the specifics of that error in the error message.

File: lib/puppet/provider/reboot_notify/notify.rb [hunk 3]
```diff
@@ -20,11 +20,15 @@ def exists?
   end
 
   def create
-    File.open(@target,'w'){|fh| fh.puts(PSON.pretty_generate({}))}
+    begin
+      File.open(@target,'w'){|fh| fh.puts(JSON.pretty_generate({}))}
+    rescue
```
Comment (lnemsick-simp) [inline lib/puppet/provider/reboot_notify/notify.rb]: I'm not a fan of blanket rescues and would prefer to see rescue of specific error types or their base class, and a reporting of the error they provide.  For example,

rescue Errno::ENOENT, Errno::EACCES, JSON::JSONError, ArgumentError, TypeError => e
  *OR*
rescue SystemCallError, JSON::JSONError => e
  raise(Puppet::Error, "reboot_notify:  Could not create '#{@target}' - #{e.message}")

File: lib/puppet/provider/reboot_notify/notify.rb [hunk 4]
```diff
@@ -11,7 +11,7 @@ def exists?
     @records = {}
 
     begin
-      @records = PSON.parse(File.read(@target))
+      @records = JSON.parse(File.read(@target))
     rescue
```
Comment (lnemsick-simp) [inline lib/puppet/provider/reboot_notify/notify.rb]: Would rather see rescue of exceptions raised by JSON.parse() and File.read(), instead of blanket exception.  For example,

raise JSON::JSONError,ArgumentError,TypeError,SystemCallError

File: lib/puppet/provider/reboot_notify/notify.rb [hunk 5]
```diff
@@ -45,9 +53,9 @@ def self.post_resource_eval
     target = "#{Puppet[:vardir]}/reboot_notifications.json"
     records = {}
     begin
-      records = PSON.parse(File.read(target))
+      records = JSON.parse(File.read(target))
     rescue
```
Comment (lnemsick-simp) [inline lib/puppet/provider/reboot_notify/notify.rb]: Would rather see rescue of exceptions raised by File.read() and JSON.parse() and reporting of the error these exceptions provide. . For example,

raise JSON::JSONError,ArgumentError,TypeError,SystemCallError => e
   raise(Puppet::Error, "Could not parse file: #{target} - #{e.message}")

File: lib/puppet/provider/reboot_notify/notify.rb [hunk 6]
```diff
@@ -57,18 +65,22 @@ def self.post_resource_eval
...
     }
 
-    if not records.empty? then
+    unless records.empty?
       msg = ["System Reboot Required Because:"]
 
       records.each_pair do |k,v|
         # This is a fail safe for empty 'reasons'
-        records[k]['reason'] = 'modified' if ( records[k]['reason'].nil? or records[k]['reason'].empty? )
+        records[k]['reason'] = 'modified' if ( records[k]['reason'].nil? || records[k]['reason'].empty? )
```
Comment (lnemsick-simp) [inline lib/puppet/provider/reboot_notify/notify.rb]: Why the change from 'or' to '||'?  They are equivalent and the first way is more understandable.

File: spec/unit/puppet/provider/reboot_notify/notify_spec.rb [hunk 7]
```diff
@@ -0,0 +1,170 @@
+require 'spec_helper'
```
Comment (lnemsick-simp) [inline spec/unit/puppet/provider/reboot_notify/notify_spec.rb]: Excellent test

File: lib/puppet/provider/reboot_notify/notify.rb (hunk 6 above)
Comment (trevor-vaughan) [reply lib/puppet/provider/reboot_notify/notify.rb]: Consistence with the rest of the codebase. We ended up moving to `||` from `or` across the board in the past in theory due to execution prioritization ordering.
